### PR TITLE
Added support for ground plane calibration

### PR DESCRIPTION
**Pull request by narora1**

### CODE REVIEW

**Comment by narora1** [comment]
Don't review till I run a linter

File: robot_calibration/CMakeLists.txt [hunk 1]
```diff
@@ -76,6 +76,7 @@ add_library(robot_calibration src/calibration_offset_parser.cpp
                               src/checkerboard_finder.cpp
                               src/feature_finder.cpp
                               src/led_finder.cpp
+			      src/ground_plane_finder.cpp
```
Comment (mikeferguson) [inline robot_calibration/CMakeLists.txt]: linter won't catch the indent being off here by 1

File: robot_calibration/include/robot_calibration/capture/ground_plane_finder.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ROBOT_CALIBRATION_CAPTURE_GROUND_PLANE_FINDER_H
+#define ROBOT_CALIBRATION_CAPTURE_GROUND_PLANE_FINDER_H
+
+#include <ros/ros.h>
+#include <pcl/point_types.h>
+#include <pcl_ros/point_cloud.h>
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/capture/ground_plane_finder.h]: we can't include these two files (this line, and above it) as we aren't depending on PCL anymore

File: robot_calibration/include/robot_calibration/ceres/ground_plane_error.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    std::vector<geometry_msgs::PointStamped> camera_pts =
+        camera_model_->project(data_, *offsets_);
+   
+    //ground height is zero w.r.t. the base_link
+    double z=0;
+     
+    // Compute residuals
+    for (size_t i = 0; i < camera_pts.size() ; ++i)
+    {
+      residuals[i] = camera_pts[i].point.z - z; // if camera_pts is in base frame
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/ground_plane_error.h]: this should actually be -z_ (use the param we define)

File: robot_calibration/include/robot_calibration/ceres/ground_plane_error.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  { 
+    // Update calibration offsets based on free params
+    offsets_->update(free_params[0]);
+
+    // Project the camera observations
+    std::vector<geometry_msgs::PointStamped> camera_pts =
+        camera_model_->project(data_, *offsets_);
+   
+    //ground height is zero w.r.t. the base_link
+    double z=0;
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/ground_plane_error.h]: unneeded when we use z_

File: robot_calibration/include/robot_calibration/ceres/ground_plane_error.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace robot_calibration
+{
+
+struct GroundPlaneError
+{
+
+  GroundPlaneError(Camera3dModel* camera_model, 
+                   CalibrationOffsetParser* offsets,
+                   robot_calibration_msgs::CalibrationData& data, double z)
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/ground_plane_error.h]: double z should be on the next line

File: robot_calibration/include/robot_calibration/ceres/optimizer.h [hunk 6]
```diff
@@ -33,7 +33,7 @@
 #include <robot_calibration/ceres/outrageous_error.h>
 #include <robot_calibration/models/camera3d.h>
 #include <robot_calibration/models/chain.h>
-
+#include <robot_calibration/ceres/ground_plane_error.h>
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/optimizer.h]: move this above with the other error models

File: robot_calibration/scripts/capture_poses_old [hunk 7]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
```
Comment (mikeferguson) [inline robot_calibration/scripts/capture_poses_old]: this is just a backup of the file? if so, remove

File: robot_calibration/scripts/feature.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import rosbag
```
Comment (mikeferguson) [inline robot_calibration/scripts/feature.py]: add copyright block to this (and next file in the PR)

File: robot_calibration/src/ground_plane_finder.cpp [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+  if(!waitForCloud())
+  {
+    ROS_ERROR("No point cloud data");
+    return false;
+  }
+
+  int points_total = 80;
+ 
+  std::vector<cv::Point2f> points;
+  //points.resize(points_x * points_y);
```
Comment (mikeferguson) [inline robot_calibration/src/ground_plane_finder.cpp]: remove

File: robot_calibration/src/ground_plane_finder.cpp [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+  points.resize(1 * points_total);
```
Comment (mikeferguson) [inline robot_calibration/src/ground_plane_finder.cpp]: why 1 * points_total? (this is in several places)

File: robot_calibration/src/optimizer.cpp [hunk 11]
```diff
@@ -31,7 +31,7 @@
 #include <robot_calibration/ceres/outrageous_error.h>
 #include <robot_calibration/models/camera3d.h>
 #include <robot_calibration/models/chain.h>
-
+#include <robot_calibration/ceres/ground_plane_error.h>
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: included in the header file, not needed here as well

File: robot_calibration/src/optimizer.cpp [hunk 12]
```diff
@@ -92,12 +92,14 @@ int Optimizer::optimize(OptimizationParams& params,
...
       models_[params.models[i].name] = model;
     }
-    else
+    else 
     {
       // ERROR unknown
     }
   }
 
+  //std::cout<<"data :"<<  data[1].observations[0].features[0].point.z << std::endl ;
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: remove this and the added line below

File: robot_calibration/src/optimizer.cpp [hunk 13]
```diff
@@ -120,7 +122,9 @@ int Optimizer::optimize(OptimizationParams& params,
   for (int i = 0; i < offsets_->size(); ++i)
     free_params[i] = 0.0;
 
+  double z_ = 0;
   // Houston, we have a problem...
+
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: remove newline

File: robot_calibration/src/optimizer.cpp [hunk 14]
```diff
@@ -165,11 +171,47 @@ int Optimizer::optimize(OptimizationParams& params,
...
+          std::cout << std::endl << "  z: ";
+          for (size_t k = 0; k < data[i].observations[0].features.size(); ++k)
+            std::cout << "  " << std::setw(10) << std::fixed << residuals[(k)];
+          std::cout << std::endl << std::endl;
+        }
 
         problem->AddResidualBlock(cost,
                                    NULL /* squared loss */,
-                                   free_params);
+                                   free_params_);
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: i think this is an unintentional change during rebase

File: robot_calibration/src/optimizer.cpp [hunk 15]
```diff
@@ -165,11 +171,47 @@ int Optimizer::optimize(OptimizationParams& params,
...
       }
+
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: remove newline

File: robot_calibration/src/optimizer.cpp [hunk 16]
```diff
@@ -165,11 +171,47 @@ int Optimizer::optimize(OptimizationParams& params,
             std::cout << "  " << std::setw(10) << std::fixed << residuals[(3*k + 2)];
           std::cout << std::endl << std::endl;
         }
+      
+        problem_->AddResidualBlock(cost,
+                                   NULL /* squared loss */,
+                                   free_params_);
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: i think this is supposed to be free_params now -- no underscore

**Comment by mikeferguson** [comment]
@narora1 there are also still two bagfiles in scripts that should not be there.

**Comment by mikeferguson** [comment]
Actually, its more than 2 -- there are a number of files in scripts that need to be removed

**Comment by narora1** [comment]
yup. its all clean now. Also capture_poses now records the new message type instead of just JointStates

**Comment by mikeferguson** [comment]
2 files left to remove:

 * robot_calibration/src/.models.cpp.swp
 * robot_calibration/scripts/calibration_poses.bag

After that, let's squash and we're ready to merge

**Comment by narora1** [comment]
done!

**Pull request merged**